File: 2025-12/src/day_09/boj_1644/Main.cpp
#include <bits/stdc++.h>
using namespace std;

bool isPrime[4000001];
vector<int> primes;

void sieve(int n) {
    fill(isPrime, isPrime + n + 1, true);
    isPrime[0] = isPrime[1] = false;

    for (int i = 2; i * i <= n; i++) {
        if (isPrime[i]) {
            for (int j = i * i; j <= n; j += i) {
                isPrime[j] = false;
            }
        }
    }

    for (int i = 2; i <= n; i++) {
        if (isPrime[i]) primes.push_back(i);
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    if (n == 1) {
        cout << 0;
        return 0;
    }

    sieve(n);
    int l = 0;
    int r = 0;
    int sum = 0;
    int cnt = 0;

    while (true) {
        if (sum < n) {
            sum += primes[r++];
        } else {
            if (sum == n) cnt++;
            sum -= primes[l++];
        }

        if (r == primes.size() && sum < n) break;
    }

    cout << cnt;
}

File: 2025-12/src/day_09/boj_1644/Main.java
package day_09.boj_1644;

import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        if (n == 1) {
            System.out.println(0);
            return;
        }

        List<Integer> primes = getPrimes(sieve(n));
        int left = 0;
        int right = 0;
        int sum = 0;
        int cnt = 0;

        while (true) {
            if (sum < n) {
                sum += primes.get(right++);
            } else {
                if (sum == n) cnt++;
                sum -= primes.get(left++);
            }

            if (right == primes.size() && sum < n) break;
        }

        System.out.println(cnt);
    }

    static boolean[] sieve(int n) {
        boolean[] isPrime = new boolean[1 + n];
        Arrays.fill(isPrime, true);
        isPrime[0] = isPrime[1] = false;

        for (int i = 2; i * i <= n; i++) {
            if (isPrime[i]) {
                for (int j = i * i; j <= n; j += i) {
                    isPrime[j] = false;
                }
            }
        }

        return isPrime;
    }

    static List<Integer> getPrimes(boolean[] isPrime) {
        List<Integer> primes = new ArrayList<>();

        for (int i = 2; i < isPrime.length; i++) {
            if (isPrime[i]) primes.add(i);
        }

        return primes;
    }
}
